### DISCUSSION TITLE
Seeed Studio Round Display with WaveShare RP2040-Zero

### DISCUSSION THREAD

**Discussion by ab-tools** [Bodmer/TFT_eSPI#2821] [Compatible displays and setup files]
Hello,

I'm trying to get the Seeed Studio Round Display (https://www.seeedstudio.com/Seeed-Studio-Round-Display-for-XIAO-p-5638.html) working with the WaveShare RP2040-Zero (https://www.seeedstudio.com/Seeed-Studio-Round-Display-for-XIAO-p-5638.html).

To ensure the display and my general Arduino setup is working correctly, I've tested the Seeed display first with the Seeed XIAO RP2040 (https://wiki.seeedstudio.com/XIAO-RP2040/) base board which can be directly connected to the display:
Using the corresponding user setup (https://github.com/Bodmer/TFT_eSPI/blob/master/User_Setups/Setup66_Seeed_XIAO_Round.h) this is working perfectly!

For reference, that's the PIN connection layout of the directly connected Seeed display to the Seeed base board:
![image](https://github.com/Bodmer/TFT_eSPI/assets/1347603/71cf2504-fc9c-42b6-8bd1-aa8772624bb0)

While this is working great, unfortuantely, this Seeed XIAO RP2040 base board does not offer many additional IO ports (you can basically connect the display and that's it) and is therefore not usable for my application. That's why I'm trying to get the display working with a WaveShare RP2040-Zero which offers several additional IO ports.

It is obvious that this WaveShare product cannot be "just plugged" onto the display as the PIN layout is completely different:
![image](https://github.com/Bodmer/TFT_eSPI/assets/1347603/83b4480e-c68a-4c18-ad3e-24d2e7112091)

Especially the GPIO ports `GP26`-`GP29` would need to be used for the SPI interface and this seems not to be possible (at least not without using something like SoftSPI or so). In case somebody thinks this might still work, please advice, as this would be ideal, of course.

But it should work in my opinion without problems if PINs are connected to the same GPIO ports as it is the case for the Seeed product. So instead of plugging the WaveShare base board directly onto the Seeed display, I connect the PINs manually like this:
![image](https://github.com/Bodmer/TFT_eSPI/assets/1347603/7569b6e4-0150-4c82-b39d-f6dbef8bfe9e)

When I run now the same code (including same user setup) as I do for the Seeed XIAO RP2040 base board the TFT screen backlight does turns on, but it does not show any output.

I'm still a bit new to this, so I would highly appreciate your opinion what might be wrong here.

I do think that the RP2040 PINs used should now be the same as for the Seeed XIAO RP2040 base board, so I did not change the user config, but perhaps there is still something that needs to be changed to make it work with the WaveShare RP2040-Zero board?

Again, any idea about this would be highly appreciated!

Thanks a lot
Andreas

**Comment by Bodmer** [comment]
I suspect the problem is due to the way the pins are specified in the setup file. Also make sure you select the correct RP2040 board in the IDE.  It is best to use GPIO numbers rather than Dx because different board suppliers map different GPIO to each Dx reference!

I think these GPIO pins in the setup should work with the Waveshare board if they are wired to the TFT display pins as follows:
```
#define TFT_SPI_PORT 0 // Set to 0 if SPI0 pins are used, or 1 if SPI1 pins used

#define TFT_MISO 0
#define TFT_CS   1
#define TFT_SCLK 2
#define TFT_MOSI 3
#define TFT_DC   4
#define TFT_BL   5
#define TFT_RST  6
```

**Comment by ab-tools** [comment]
First thanks a lot, @Bodmer, for taking the time to reply - even so quickly!

The PIN IDs you mentioned were not correct, but based on your tip that the Dx might be very different (should have thought about that myself!), I checked the ones from the Seeed board and they ARE very different:
https://github.com/earlephilhower/arduino-pico/blob/master/variants/seeed_xiao_rp2040/pins_arduino.h

When I just use the "raw" GPIO IDs it does work when connecting the WaveShare board to the exact same PINs using individual cables:
```
#define TFT_SCLK 2
#define TFT_MISO 4
#define TFT_MOSI 3
#define TFT_CS   27
#define TFT_DC   29
#define TFT_BL   0
#define TFT_RST -1 // as it was undefined for Seeed as well
```

Only the touch screen driver still spits out very weird touch points so far:
```
[Info]  (14.595, +1)     indev_proc_press: pressed at x:15376 y:8193    (in lv_indev.c line #819)
[Warn]  (14.628, +33)    indev_pointer_proc: X is 15376 which is greater than hor. res  (in lv_indev.c line #362)
[Warn]  (14.629, +1)     indev_pointer_proc: Y is 8193 which is greater than ver. res   (in lv_indev.c line #368)
```
But I did change the GPIO PINs for the touch driver as well, not sure what's going wrong there...
It does get the touch event, but with this very weird "touch location".

But I guess the touch part is beyond the scope of your library here, right?

Thanks again for your support
Andreas

**Comment by ab-tools** [comment]
Just wanted to thank you again for your support, @Bodmer, I can confirm all is working now:
For the touch input I just also needed to adjust the I²C PINs. Now touch is working, too. :-)

Just out of curiosity, did you ever try using SoftSPI to connect such a TFT display?

Still thinking if there would be a way to just plug the WaveShare board directly to the Seeed Screen, but this would require using `GP27` - `GP29` as SPI ports and as I understand this is not supported by the hardware SPI.

Do you think it's worth trying something like this?
https://github.com/MajenkoLibraries/SoftSPI
Or is the RP2040 too slow anyway to emulate SPI properly software-wise?

Thanks again
Andreas
